DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/GroupDelegatesController.cs

```diff
@@ -0,0 +1,217 @@
+        {
+            var delegateUser = userService.GetDelegateUserById(delegateId);
+            if (delegateUser == null)
+            {
+                return NotFound();
+            }
+
+            var centreId = User.GetCentreId();
+            var groupName = groupsService.GetGroupName(groupId, centreId);
+
+            var model = new ConfirmDelegateAddedViewModel(delegateUser.FullName, groupName!, groupId);
+            return View(model);
+        }
+
+        [HttpGet("Add/SelectDelegate/AllItems")]
+        public IActionResult SelectDelegateAllItems(int groupId)
+        {
+            var centreId = User.GetCentreId();
+            var jobGroups = jobGroupsService.GetJobGroupsAlphabetical();
+            var customPrompts = centreCustomPromptHelper.GetCustomPromptsForCentre(centreId);
+            var delegateUsers = userService.GetDelegatesNotRegisteredForGroupByGroupId(groupId, centreId);
+
+            var model = new SelectDelegateAllItemsViewModel(
+                delegateUsers,
+                jobGroups,
+                customPrompts,
+                groupId
+            );
+
+            return View(model);
+        }
+
+
+        [HttpGet]
+        [Route("{delegateId:int}/Remove")]
+        public IActionResult RemoveGroupDelegate(int groupId, int delegateId)
+        {
+            var centreId = User.GetCentreId();
+            var groupName = groupsService.GetGroupName(groupId, centreId);
+            var groupDelegates = groupsService.GetGroupDelegates(groupId).ToList();
+            var delegateUser = groupDelegates.SingleOrDefault(gd => gd.DelegateId == delegateId);
+
+            if (groupName == null || delegateUser == null)
+            {
+                return NotFound();
+            }
+
+            var progressId = groupsService.GetRelatedProgressIdForGroupDelegate(groupId, delegateId);
+
+            var model = new RemoveGroupDelegateViewModel(delegateUser, groupName, groupId, progressId);
+
+            return View(model);
+        }
+
+        [HttpPost]
+        [Route("{delegateId:int}/Remove")]
+        public IActionResult RemoveGroupDelegate(RemoveGroupDelegateViewModel model, int groupId, int delegateId)
+        {
+            var centreId = User.GetCentreId();
+            var groupName = groupsService.GetGroupName(groupId, centreId);
+            var groupDelegates = groupsService.GetGroupDelegates(groupId).ToList();
+            var delegateUser = groupDelegates.SingleOrDefault(gd => gd.DelegateId == delegateId);
+
+            if (groupName == null || delegateUser == null)
+            {
+                return NotFound();
+            }
+
+            if (!model.ConfirmRemovalFromGroup)
+            {
+                ModelState.AddModelError(
+                    nameof(RemoveGroupDelegateViewModel.ConfirmRemovalFromGroup),
+                    "You must confirm before removing this user from the group"
+                );
+                return View(model);
+            }

```

Comment:
Why is this in the controller and not on the model?

Reply:
Good question. This was just old code moved to a new place but worth the refactor.

Reply:
Done.